packages/cosmic-swingset/src/launch-chain.js

```diff
@@ -733,6 +742,63 @@ export async function launch({
         return true;
       }
 
+      case ActionType.ENACTED_UPGRADE: {
+        // Install and execute new core proposals.
+        const {
+          upgradePlan: { info: upgradeInfoJson = null } = {},
+          blockHeight,
+          blockTime,
+        } = action;
+        const upgradePlanInfo = upgradeInfoJson && JSON.parse(upgradeInfoJson);
+
+        // Handle the planned core proposals as just another action.
+        const { coreProposals = [] } = upgradePlanInfo || {};
+
+        if (!coreProposals.length) {
+          // Nothing to do.
+          return undefined;
+        }
+
+        const { bundles, code: coreEvalCode } =
+          await extractCoreProposalBundles(coreProposals, vatconfig, {
+            handleToBundleSpec: async (handle, source, _sequence, _piece) => {
+              const bundle = await bundleSource(source);
+              const { endoZipBase64Sha512: hash } = bundle;
+              const bundleID = `b1-${hash}`;
+              handle.bundleID = bundleID;
+              harden(handle);
+              return harden([handle.bundleID, bundle]);
+            },
+          });
+
+        for (const bundle of Object.values(bundles)) {
+          await controller.validateAndInstallBundle(bundle);
+        }
+
+        // Now queue the code for evaluation.
+        const coreEvalAction = {
+          type: ActionType.CORE_EVAL,
+          blockHeight,
+          blockTime,
+          evals: [
+            {
+              json_permits: 'true',
+              js_code: coreEvalCode,
+            },
+          ],
+        };
+        highPriorityQueue.push({
+          context: {
+            blockHeight,
+            txHash: ActionType.ENACTED_UPGRADE,
```

Comment:
I believe vbank already has synthesized `txHash` with a specific format. Let's use that same format?

Maybe something like `x/upgrade` ?